bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java

```diff
@@ -550,11 +599,451 @@ private static RetryOptions createRetryOptions(Configuration configuration) {
+
+    String dataHostOverride =
+        configuration.get(BIGTABLE_HOST_KEY, BigtableOptions.BIGTABLE_DATA_HOST_DEFAULT);
+    int portNumber = configuration.getInt(BIGTABLE_PORT_KEY, BIGTABLE_PORT_DEFAULT);
+
+    String endpoint = dataHostOverride + ":" + portNumber;
+    LOG.debug("API Data endpoint hostname:portNumber is %s", endpoint);
+
+    stubSettings
+        .setCredentialsProvider(buildCredentialProvider(configuration))
+        .setEndpoint(endpoint)
+        .setHeaderProvider(buildHeaderProvider(configuration));
+
+    boolean usePlainTextNegotiation =
+        configuration.getBoolean(BIGTABLE_USE_PLAINTEXT_NEGOTIATION, false);
+    if (usePlainTextNegotiation) {
+      int channelCount =
+          configuration.getInt(
+              BIGTABLE_DATA_CHANNEL_COUNT_KEY, BIGTABLE_DATA_CHANNEL_COUNT_DEFAULT);
+
+      stubSettings.setTransportChannelProvider(
+          buildPlainTextChannelProvider(endpoint, channelCount));
+    }
+
+    Duration shortRpcTimeoutMs =
+        ofMillis(configuration.getInt(BIGTABLE_RPC_TIMEOUT_MS_KEY, SHORT_TIMEOUT_MS_DEFAULT));
+
+    // Configuration for rpcTimeout & totalTimeout for non-streaming operations.
+    stubSettings.checkAndMutateRowSettings().setSimpleTimeoutNoRetries(shortRpcTimeoutMs);
+
+    stubSettings.readModifyWriteRowSettings().setSimpleTimeoutNoRetries(shortRpcTimeoutMs);
+
+    Set<StatusCode.Code> retryCodes = buildRetryCodes(configuration);
+
+    buildBulkMutationsSettings(stubSettings, configuration, retryCodes);
+
+    buildReadRowsSettings(stubSettings, configuration, retryCodes);
+
+    stubSettings
+        .readRowSettings()
+        .setRetryableCodes(retryCodes)
+        .setRetrySettings(
+            buildIdempotentRetrySettings(
+                stubSettings.readRowSettings().getRetrySettings(), configuration));
+
+    stubSettings
+        .mutateRowSettings()
+        .setRetryableCodes(retryCodes)
+        .setRetrySettings(
+            buildIdempotentRetrySettings(
+                stubSettings.mutateRowSettings().getRetrySettings(), configuration));
+
+    stubSettings
+        .sampleRowKeysSettings()
+        .setRetryableCodes(retryCodes)
+        .setRetrySettings(
+            buildIdempotentRetrySettings(
+                stubSettings.sampleRowKeysSettings().getRetrySettings(), configuration));
+
+    String emulatorHost = configuration.get(BIGTABLE_EMULATOR_HOST_KEY);
+    if (!Strings.isNullOrEmpty(emulatorHost)) {
+      try {
+        int lastIndexOfCol = emulatorHost.lastIndexOf(":");
+        int port = Integer.parseInt(emulatorHost.substring(lastIndexOfCol + 1));
+
+        String hostPort = emulatorHost.substring(0, lastIndexOfCol) + ":" + port;
+        stubSettings
+            .setCredentialsProvider(NoCredentialsProvider.create())
+            .setEndpoint(hostPort)
+            .setTransportChannelProvider(
+                InstantiatingGrpcChannelProvider.newBuilder()
```

Comment:
If we can't reuse veneer's emulator support, then we should at least reuse the existing grpc channel provider and avoid duplicating these settings